src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,419 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → ideal R`and operations on
+them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+open set_like direct_sum set
+open_locale big_operators pointwise direct_sum
+
+section homogeneous_def
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+variables [decidable_eq ι] [add_comm_monoid ι]  [graded_algebra 𝒜]
```

Comment:
```suggestion
variables {ι R A : Type*} [comm_semiring R] [semiring A] [algebra R A]
variables (𝒜 : ι → submodule R A)
variables [decidable_eq ι] [add_monoid ι] [graded_algebra 𝒜]
```
works fine for me (and in the later variables too). You might need to merge master, and wait for (or adopt) #12208

Reply:
done